tests/returntypes/test_parameter_shift_new.py

```diff
@@ -0,0 +1,2471 @@
+        assert np.allclose(grad, numeric_val, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("theta", np.linspace(-2 * np.pi, np.pi, 7))
+    def test_CRot_gradient(self, theta, tol):
+        """Tests that the automatic gradient of an arbitrary controlled Euler-angle-parameterized
+        gate is correct."""
+        dev = qml.device("default.qubit", wires=2)
+        a, b, c = np.array([theta, theta**3, np.sqrt(2) * theta])
+
+        with qml.tape.QuantumTape() as tape:
+            qml.QubitStateVector(np.array([1.0, -1.0], requires_grad=False) / np.sqrt(2), wires=0)
+            qml.CRot(a, b, c, wires=[0, 1])
+            qml.expval(qml.PauliX(0))
+
+        tape.trainable_params = {1, 2, 3}
+
+        res = dev.execute_new(tape)
+        expected = -np.cos(b / 2) * np.cos(0.5 * (a + c))
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        tapes, fn = qml.gradients.param_shift(tape)
+        assert len(tapes) == 4 * len(tape.trainable_params)
+
+        grad = fn(dev.batch_execute_new(tapes))
+        expected = np.array(
+            [
+                0.5 * np.cos(b / 2) * np.sin(0.5 * (a + c)),
+                0.5 * np.sin(b / 2) * np.cos(0.5 * (a + c)),
+                0.5 * np.cos(b / 2) * np.sin(0.5 * (a + c)),
+            ]
+        )
+        assert isinstance(grad, tuple)
+        assert len(grad) == 3
+        for idx, g in enumerate(grad):
+            assert np.allclose(g, expected[idx], atol=tol, rtol=0)
+
+        # compare to finite differences
+        tapes, fn = qml.gradients.finite_diff(tape)
+        numeric_val = np.squeeze(fn(dev.batch_execute_new(tapes)))
+        for idx, g in enumerate(grad):
+            assert np.allclose(g, numeric_val[idx], atol=tol, rtol=0)
+
+    def test_gradients_agree_finite_differences(self, tol):
+        """Tests that the parameter-shift rule agrees with the first and second
+        order finite differences"""
+        params = np.array([0.1, -1.6, np.pi / 5])
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(params[0], wires=[0])
+            qml.CNOT(wires=[0, 1])
+            qml.RY(-1.6, wires=[0])
+            qml.RY(params[1], wires=[1])
+            qml.CNOT(wires=[1, 0])
+            qml.RX(params[2], wires=[0])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+
+        tape.trainable_params = {0, 2, 3}
+        dev = qml.device("default.qubit", wires=2)
+
+        grad_F1 = grad_fn(tape, dev, fn=qml.gradients.finite_diff, approx_order=1)
+        grad_F2 = grad_fn(
+            tape, dev, fn=qml.gradients.finite_diff, approx_order=2, strategy="center"
+        )
+        grad_A = grad_fn(tape, dev)
+
+        # gradients computed with different methods must agree
+        assert np.allclose(grad_A, grad_F1, atol=tol, rtol=0)
+        assert np.allclose(grad_A, grad_F2, atol=tol, rtol=0)
+
+    # TODO: remove xfail when var/finite diff works
+    @pytest.mark.xfail
```

Comment:
Is it an old comment, should we try to remove the xfail?

Reply:
It doesn't work, throws an error from finite diff logic. I'll update the comment to remove confusion.